src/icrar/leap-accelerate/algorithm/cuda/PhaseRotate.cu

```diff
@@ -149,20 +231,91 @@ namespace cuda
         }
     }
 
+    __global__ void g_RotateVisibilitiesBC(
+        cuDoubleComplex* pintegration_data, int integration_data_rows, int integration_data_cols, int integration_data_depth,
+        int integration_channels,
+        int integration_baselines,
+        int polarizations,
+        icrar::cpu::Constants constants,
+        Eigen::Matrix3d dd,
+        double2 direction,
+        double3* uvw, int uvwLength,
+        double3* oldUVW, int oldUVWLegth,
+        cuDoubleComplex* pavg_data, int avg_dataRows, int avg_dataCols)
+    {
+        using Tensor2Xcucd = Eigen::Tensor<cuDoubleComplex, 2>;
+        using Tensor3Xcucd = Eigen::Tensor<cuDoubleComplex, 3>;
+
+        //parallel execution per channel
+        int baseline = blockDim.x * blockIdx.x + threadIdx.x;
+        int channel = blockDim.y * blockIdx.y + threadIdx.y;
+
+        if(baseline < integration_baselines && channel < integration_channels)
+        {
```

Comment:
This is an old opinionated argument but I tend to find codebases with early returns harder to maintain. In many cases to write expected path you need to be aware of the preceding conditions and at most there should only be about 3 of them otherwise the function signature should be more strict or something like a strategy, fallback or chain of responsibility pattern should be considered.